proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,450 @@
+}
+```
+
+`A::foo` in the example above should be available in Kotlin as `fun foo(x: 
+String): String?`
+
+
+### Type qualifier default<a name="type-qualifier-default"></a>
+[`@TypeQualifierDefault`](https://aalmiray.github.io/jsr-305/apidocs/javax/annotation/meta/TypeQualifierDefault.html)
+allows introducing annotations that when being applied define the default
+nullability within the scope of the annotated element.
+
+To make it work new annotation class must be annotated with 
+`@TypeQualifierDefault` and `@NonNull` or any type qualifier nickname to the 
+latter.
+The single argument for the `@TypeQualifierDefault` defines a set of element 
+types for which the annotation may enhance a type. 
+
+When determining immediate nullability of a type the Kotlin compiler should look
+for nullability annotation on the type itself at first.
+- If there is one, then it should use it
+- Otherwise, nullability is determined by the innermost enclosing element 
+annotated with `TypeQualifierDefault` and having appropriate applicability 
+in the argument:
+    - `ElementType.METHOD` for return type of methods
+    - `ElementType.PARAMETER` for value parameters
+    - `ElementType.FIELD` for fields
+    - `ElementType.TYPE_USE` for any type including type arguments, upper bounds 
+    of type parameters and wildcard types
+    
+Note, that [`javax.annotation.ParametersAreNonnullByDefault`](https://aalmiray.github.io/jsr-305/apidocs/javax/annotation/ParametersAreNonnullByDefault.html) 
+annotation should work automatically by applying the rules above:
+```java
+@Nonnull
+@TypeQualifierDefault(value=PARAMETER)
+@Retention(value=RUNTIME)
+public @interface ParametersAreNonnullByDefault {}
+```
+
+Also, it's important to support reading Java package annotations because it's
+supposed to be the most common way to declare a default nullability.
+
+```java
+@Nonnull
+@TypeQualifierDefault({ElementType.METHOD, ElementType.PARAMETER})
+public @interface NonNullApi {
+}
+
+@Nonnull(when = When.MAYBE)
+@TypeQualifierDefault({ElementType.METHOD, ElementType.PARAMETER})
+public @interface NullableApi {
+}
+```
+
+```java
+// FILE: test/package-info.java
+@NonNullApi // declaring all types in package 'test' as non-nullable by default
+package test;
+```
+
+```java
+// FILE: test/A.java
+package test;
+
+@NullableApi // overriding default nullability from the package
+interface A {
+    String foo(String x); // fun foo(x: String?): String?
+ 
+    @NotNullApi // overriding default from the class
+    String bar(String x, @Nullable String y); // fun bar(x: String, y: String?): String 
+    
+    // The type of `x` parameter remains flexible because there's explicit UNKNOWN-marked
+    // nullability annotation
+    String baz(@Nonnull(when = When.UNKNOWN) String x); // fun baz(x: String!): String?
+}
+
+```
+
+### `@UnderMigration` annotation
+Current proposal suggests to add the following meta-annotation in 
+the `kotlin.annotations.jvm` package:
```

Comment:
You should probably also highlight that it's in a separate library, not in kotlin-stdlib